core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -655,52 +657,79 @@ private void doStop(CompletableFuture<Void> future,
-                            bossGroup.shutdownGracefully();
-                            bossGroup.terminationFuture().addListener(unused6 -> {
-                                if (remainingBossGroups.decrementAndGet() != 0) {
-                                    // There are more boss groups to terminate.
-                                    return;
-                                }
-
-                                // Boss groups have been terminated completely.
+                        workerShutdownFuture.addListener(unused7 -> {
+                            final Set<EventLoopGroup> bossGroups =
+                                    Server.this.serverChannels.stream()
+                                                              .map(ch -> ch.eventLoop().parent())
+                                                              .collect(toImmutableSet());
+
+                            // If started to shutdown before initializing a boss group,
+                            // complete the future immediately.
+                            if (bossGroups.isEmpty()) {
                                 finishDoStop(future);
+                                return;
+                            }
+
+                            // Shut down all boss groups and wait until they are terminated.
+                            final AtomicInteger remainingBossGroups = new AtomicInteger(bossGroups.size());
+                            bossGroups.forEach(bossGroup -> {
+                                bossGroup.shutdownGracefully();
+                                bossGroup.terminationFuture().addListener(unused8 -> {
+                                    if (remainingBossGroups.decrementAndGet() != 0) {
+                                        // There are more boss groups to terminate.
+                                        return;
+                                    }
+
+                                    // Boss groups have been terminated completely.
+                                    finishDoStop(future);
+                                });
                             });
                         });
-                    });
 
+                        return null;
+                    });
                     return null;
                 });
 
                 return null;
             });
         }
 
+        private CompletableFuture<Void> shutdownServerHandlers() {
+            if (config.gracefulShutdown().timeout().isZero()) {
+                return UnmodifiableFuture.completedFuture(null);
+            }
+
+            final Set<Channel> children = connectionLimitingHandler.children();
+            final List<CompletableFuture<Void>> closeFutures = new ArrayList<>(children.size());
+            for (Channel ch : children) {
+                final HttpServerHandler serverHandler = ch.pipeline().get(HttpServerHandler.class);
+                if (serverHandler != null) {
+                    closeFutures.add(serverHandler.shutdown(ch));
```

Comment:
This is called from the `startStopExecutor`, which breaks the assumption that `HttpServerHandler#cleanup` is called from the event loop assigned to the channel.

Is this intentional?

Reply:
Nice catch, it was my mistake.